File: src/java/foundgine-core/src/main/java/com/foundgine/core/abstractions/SemanticIdentity.java
package com.foundgine.core.abstractions;

import java.nio.charset.StandardCharsets;

/**
 * Port of {@code Foundgine.Core.Abstractions.SemanticIdentity}.
 *
 * <p>
 * Canonical semantic identity and stable hashing rules shared by runtime
 * identifiers and the AOT generator. The canonical key is intentionally
 * independent of JVM/CLR metadata, declaration order, and runtime state, so the
 * 64-bit hash produced here is byte-for-byte identical to the one produced by
 * the original C# implementation for the same input string.
 */
public final class SemanticIdentity {

	public static final String ENTITY_NAMESPACE = "entity";
	public static final String FIELD_NAMESPACE = "field";
	public static final String RELATIONSHIP_NAMESPACE = "relationship";
	public static final String TABLE_NAMESPACE = "table";
	public static final String COLUMN_NAMESPACE = "column";
	public static final String MODEL_NAMESPACE = "model";
	public static final String CONNECTION_NAMESPACE = "connection";
	public static final String AUTHORIZATION_NAMESPACE = "authorization";

	private static final long FNV_OFFSET_BASIS = 0xcbf29ce484222325L; // 14695981039346656037 (unsigned)
	private static final long FNV_PRIME = 0x100000001b3L; // 1099511628211

	private SemanticIdentity() {
	}

	public static String entityKey(String semanticName) {
		return key(ENTITY_NAMESPACE, semanticName);
	}

	public static String fieldKey(String semanticEntityName, String semanticFieldName) {
		return key(FIELD_NAMESPACE, pair(semanticEntityName, semanticFieldName));
	}

	public static String relationshipKey(String semanticEntityName, String semanticRelationshipName) {
		return key(RELATIONSHIP_NAMESPACE, pair(semanticEntityName, semanticRelationshipName));
	}

	public static String tableKey(String storageName) {
		return key(TABLE_NAMESPACE, storageName);
	}

	public static String columnKey(String storageName, String columnName) {
		return key(COLUMN_NAMESPACE, pair(storageName, columnName));
	}

	public static String modelKey(String semanticName) {
		return key(MODEL_NAMESPACE, semanticName);
	}

	public static String connectionKey(String semanticModelName, String semanticConnectionName) {
		return key(CONNECTION_NAMESPACE, pair(semanticModelName, semanticConnectionName));
	}

	public static String authorizationKey(String declaringType, String authorizationName) {
		return key(AUTHORIZATION_NAMESPACE, pair(declaringType, authorizationName));
	}

	/**
	 * Computes the Foundgine stable 64-bit identity hash (FNV-1a over UTF-8 bytes).
	 */
	public static long hash(String canonicalKey) {
		if (canonicalKey == null || canonicalKey.isBlank()) {
			throw new IllegalArgumentException("Canonical identity key is required.");
		}

		long hash = FNV_OFFSET_BASIS;
		for (byte b : canonicalKey.getBytes(StandardCharsets.UTF_8)) {
			hash ^= (b & 0xFFL);
			hash *= FNV_PRIME;
		}

		// Zero is reserved as an invalid/unassigned identity.
		return hash == 0L ? 1L : hash;
	}

	public static String normalize(String value, String parameterName) {
		if (value == null || value.isBlank()) {
			throw new IllegalArgumentException("Identity component is required: " + parameterName);
		}
		return value.trim();
	}

	/** {@code value} is interpreted as an unsigned 64-bit identity. */
	public static long validateExplicitId(long value, String description) {
		if (value == 0L) {
			throw new IllegalArgumentException(
					"Explicit " + description + " identity 0 is reserved and cannot be assigned.");
		}
		return value;
	}

	private static String key(String namespace, String value) {
		return namespace + ":" + normalize(value, "value");
	}

	private static String pair(String left, String right) {
		return normalize(left, "left") + "." + normalize(right, "right");
	}
}

File: src/csharp/Foundgine.Core/Interfaces/SemanticIdentity.cs
using System.Text;

namespace Foundgine.Core.Abstractions;

/// <summary>
/// Canonical semantic identity and stable hashing rules shared by runtime
/// identifiers and the AOT generator. The canonical key is intentionally
/// independent of CLR metadata tokens, declaration order, and runtime state.
/// </summary>
public static class SemanticIdentity
{
    public const string EntityNamespace = "entity";
    public const string FieldNamespace = "field";
    public const string RelationshipNamespace = "relationship";
    public const string TableNamespace = "table";
    public const string ColumnNamespace = "column";
    public const string ModelNamespace = "model";
    public const string ConnectionNamespace = "connection";
    public const string AuthorizationNamespace = "authorization";

    public static string EntityKey(string semanticName) =>
        Key(EntityNamespace, semanticName);

    public static string FieldKey(string semanticEntityName, string semanticFieldName) =>
        Key(FieldNamespace, Pair(semanticEntityName, semanticFieldName));

    public static string RelationshipKey(string semanticEntityName, string semanticRelationshipName) =>
        Key(RelationshipNamespace, Pair(semanticEntityName, semanticRelationshipName));

    public static string TableKey(string storageName) =>
        Key(TableNamespace, storageName);

    public static string ColumnKey(string storageName, string columnName) =>
        Key(ColumnNamespace, Pair(storageName, columnName));

    public static string ModelKey(string semanticName) =>
        Key(ModelNamespace, semanticName);

    public static string ConnectionKey(string semanticModelName, string semanticConnectionName) =>
        Key(ConnectionNamespace, Pair(semanticModelName, semanticConnectionName));

    public static string AuthorizationKey(string declaringType, string authorizationName) =>
        Key(AuthorizationNamespace, Pair(declaringType, authorizationName));

    /// <summary>Computes the Foundgine stable 64-bit identity hash.</summary>
    public static ulong Hash(string canonicalKey)
    {
        if (string.IsNullOrWhiteSpace(canonicalKey))
            throw new ArgumentException("Canonical identity key is required.", nameof(canonicalKey));

        const ulong offsetBasis = 14695981039346656037UL;
        const ulong prime = 1099511628211UL;
        var hash = offsetBasis;

        foreach (var b in Encoding.UTF8.GetBytes(canonicalKey))
        {
            hash ^= b;
            hash *= prime;
        }

        // Zero is reserved as an invalid/unassigned identity.
        return hash == 0 ? 1UL : hash;
    }

    public static string Normalize(string value, string parameterName)
    {
        if (string.IsNullOrWhiteSpace(value))
            throw new ArgumentException("Identity component is required.", parameterName);

        return value.Trim();
    }

    public static ulong ValidateExplicitId(ulong value, string description) =>
        value == 0
            ? throw new ArgumentOutOfRangeException(nameof(value), value,
                $"Explicit {description} identity 0 is reserved and cannot be assigned.")
            : value;

    private static string Key(string @namespace, string value) =>
        @namespace + ":" + Normalize(value, nameof(value));

    private static string Pair(string left, string right) =>
        Normalize(left, nameof(left)) + "." + Normalize(right, nameof(right));
}
